tutoraspects/templates/aspects/apps/superset/pythonpath/delete_assets.py

```diff
@@ -0,0 +1,68 @@
+"""Delete all unused Aspects assets from Superset tables"""
+import os
+
+import logging
+import yaml
+from collections import defaultdict
+from flask import g
+
+from superset import security_manager
+from superset.extensions import db
+from superset.models.slice import Slice
+from superset.connectors.sqla.models import SqlaTable
+from superset.tags.models import TaggedObject, ObjectType
+from superset.commands.chart.delete import DeleteChartCommand
+from superset.commands.dataset.delete import DeleteDatasetCommand
+from superset.commands.tag.delete import DeleteTaggedObjectCommand
+from sqlalchemy.exc import NoResultFound
+from superset.commands.exceptions import CommandInvalidError
+
+
+logger = logging.getLogger("delete_assets")
+PYTHONPATH = "/app/pythonpath"
+
+ASSET_TABLES = {'charts': Slice, 'datasets': SqlaTable}
+ASSET_COMMANDS = {'charts': DeleteChartCommand, 'datasets': DeleteDatasetCommand}
+OBJECT_TYPES = {'charts': ObjectType.chart, 'datasets': ObjectType.dataset} 
+
+def delete_assets(unused_uuids,asset_type):
+    """Delete unused assets and their translated versions"""
+    with open(os.path.join(PYTHONPATH,"translated_asset_mapping.yaml"),'r', encoding="utf-8") as file:
+        mapping = yaml.safe_load_all(file)
+        for line in mapping:
+            translated_ids = line
+
+    id_list = set()
+    uuid_list = set()
+    for uuid in unused_uuids:
+        try:
+            row = db.session.query(ASSET_TABLES[asset_type]).filter_by(uuid=uuid).one()
+            id_list.add(row.id)
+            uuid_list.add(uuid)
+
+            if uuid in translated_ids:
+                for child_uuid in translated_ids[uuid]:
+                    row = db.session.query(ASSET_TABLES[asset_type]).filter_by(uuid=child_uuid).one()
+                    id_list.add(row.id)
+                    uuid_list.add(child_uuid)
+        except NoResultFound:
+            continue
+        
+    if len(id_list) > 0:
+        logger.warning(f'Deleting the following {asset_type}: ')
+        logger.warning(uuid_list)
+
+        # Force our use to the admin user to prevent errors on delete
+        g.user = security_manager.find_user(username="{{SUPERSET_ADMIN_USERNAME}}")
+
+        try:
+            # Delete tagged object rows first
+            for id in id_list:
+                tag_command = DeleteTaggedObjectCommand(OBJECT_TYPES[asset_type], id)
+                tag_command.run()
```

Comment:
This is not necessary as per Superset 4.0.2, and on Superset 4.1.0 this should happen automatically by the backend

Reply:
It does happen automatically, but if there are multiple tag rows per object (which there are locally at least) then it fails. I created an issue in the superset repo